packages/nextjs/guides/blockchain-randomness-solidity.md

```diff
@@ -0,0 +1,357 @@
+
+This fundamental constraint means any randomness derived from on-chain data is inherently **public and predictable**. Understanding this limitation is critical for building secure decentralized applications, especially blockchain games, NFT drops, and lottery systems.
+
+**The PRNG Problem:** Traditional computers use pseudorandom number generators (PRNGs) seeded with high-entropy sources like system clocks or hardware noise. Smart contracts can't access these node-specific values without breaking consensus, forcing reliance on public blockchain data.
+
+### The Determinism Problem
+
+Ethereum's strength is its predictability: every node executing the same transaction must arrive at identical results. This makes **on-chain randomness** an oxymoronany "random" value computable by one node can be computed by all nodes, including attackers.
+
+---
+
+## 2. Common Pitfalls: Why Block Variables Fail
+
+Developers often turn to seemingly unpredictable block variables for randomness. **This is always a mistake.**
+
+### Block.timestamp and Blockhash
+
+```solidity
+// NEVER DO THIS - Vulnerable to manipulation
+function badRandom() public view returns (uint256) {
+    return uint256(keccak256(abi.encodePacked(
+        blockhash(block.number - 1),
+        block.timestamp,
+        msg.sender
+    )));
+}
+```
+
+**Why this fails:**
+
+- **Validator Control:** Block proposers can manipulate timestamps within bounds and choose which transactions to include.
+- **Predictable Attacks:** Any user can compute the same "random" value and only execute transactions with favorable outcomes.
+- **Blockhash Limitations:** Only accessible for 256 recent blocks; older queries return zero.
+- **256-Block Exploit:** If contract logic can be delayed for >256 blocks, blockhash() returns 0x0, creating an exploitable known value.
+- **Validator Re-rolling:** Validators can discard unfavorable blocks and try again if the value at stake exceeds block rewards.
+
+### Block.prevrandao (The RANDAO Trap)
+
+After Ethereum's Proof-of-Stake transition, `block.difficulty` became `block.prevrandao`. While better than timestamp, it's still **not secure for smart contracts**.
+
+**What is RANDAO?** It's a mechanism where validators contribute to a shared random number through cryptographic commitments. However, this doesn't solve the manipulation problem.
+
+```solidity
+// Still vulnerable - Don't use for valuable applications
+function stillBad() public view returns (uint256) {
+    return uint256(keccak256(abi.encodePacked(block.prevrandao, msg.sender)));
+}
+```
+
+**Vulnerabilities:**
+
+- **Validator Bias:** The proposing validator knows prevrandao in advance and can choose to skip their slot if the outcome is unfavorable.
+- **Same-Block Predictability:** All transactions in a block see the same prevrandao value.
+
```

Comment:
We can add here that the way to avoid this is using a future randao, and add a link to the SE-2 extension https://speedrunethereum.com/builds/1449d6d5-e014-531c-91f6-abb747281375